### PR TITLE
Br wads fix 20210413 1428

### PR DESCRIPTION
**Pull request by alanlhc**
Summary of changes
==============
1) iotconnect.c
- Separate IotConnectSdk_Init() to IotConnectSdk_Init() and IotConnectSdk_Connect().
    - IotConnectSdk_Init() => Perform discover only.
    - IotConnectSdk_Connect() => Perform connection to MQTT broker.
 
- Added IotConnectSdk_Abort() function. iotc_nrf_mqtt_abort() is invoked in this function to disconnect MQTT broker locally.
- Modify IotConnectSdk_SendPacket() to add a unsigned long pointer to receive the message Id. Main application can use this message id to track the message sent when acknowledge timeout event occurred.

2) iotconnect.h
- Added declaration for IotConnectSdk_Connect() and IotConnectSdk_Abort().
- Modify IotConnectSdk_SendPacket() to take in unsigned long pointer as 2nd parameter.
- Modify IOTCONNECT_CLIENT_CONFIG structure to add a callback pointer for message acknowledge timeout event.

3) iotconnect_mqtt.c
- Added iotc_nrf_mqtt_abort() to invoke mqtt_abort() to disconnect MQTT broker locally.
- Added helper functions to keep track of message acknowledge timeout and trigger callback function when no PUBACK received for the specific message id after 5 seconds.

4) iotconnect_mqtt.h
- Added declaration for iotc_nrf_mqtt_abort().

5) main.c from iotc-basic
- Modified to have the following features:
    - montior LTE registration and cell status.
        - When detected LTE link down, IotConnectSdk_Abort() is invoked to disconnect the MQTT broker locally.
        - When detected LTE resume, IotConnectSdk_Connect() is invoked to re-connect back to MQTT broker and resume sending of telemetry data.
....- Added message acknowledge timeout event handler to print out acknowledge time message for which message id.

6) main.c from iotc-sensors-gps
- Modified the sdk_run() to connect to MQTT broker using IotConnectSdk_Connect().
- Modified IotConnectSdk_SendPacket() to have 2nd parameter as NULL.

### CODE REVIEW

**Review by nik-markovic — COMMENTED** [review]
Comment blocks that don't add value like "Function to find msg ID from the list." should be removed because the function name msg_wait_puback_list_find makes is obvious that it's supposed to do.

**Review by nik-markovic — COMMENTED** [review]
Would it be too much work or not desirable to move the puback timeout communication into the status message IOT_CONNECT_STATUS_CB?

File: iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c [hunk 1]
```diff
@@ -50,10 +69,141 @@ static int fds_init(struct mqtt_client *c);
...
+
+        } else {
+            p = p->p_next;
+        }
+    }
+}
 
 static unsigned int get_next_message_id() {
+
     if (rolling_message_id + 1 >= UINT_MAX) {
```
Comment (nik-markovic) [inline iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c:205]: message ID is uint32_t and UINT_MAX is max for unsigned int. I think that the max should be checked against UINT32_MAX (if it exists), or 0xFFFFFFFF, or the message id type is changed to unsigned int. Personally, I would prefer unsigned int for compatibility, but it's up to you.

File: iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c [hunk 2]
```diff
@@ -438,3 +617,17 @@ void iotc_nrf_mqtt_loop(void) {
...
 }
+
+void iotc_nrf_mqtt_disconnect() {
+
+    mqtt_disconnect(&client);
+    //clear list.
+    msg_wait_puback_list_clear();
+}
+
+void iotc_nrf_mqtt_abort() {
```
Comment (nik-markovic) [inline iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c:628]: What is the purpose of the abort if disconnect does the same?

**Comment by alanlhc** [comment]
> Would it be too much work or not desirable to move the puback timeout communication into the status message IOT_CONNECT_STATUS_CB?

Hi Nik,
From the naming of "IOT_CONNECT_STATUS_CB" it is clearly defined as messages for "connect status". It would be inappropriate to add "Message acknowledge timeout" message  to IOT_CONNECT_STATUS_CB.

**Comment by alanlhc** [comment]
> Would it be too much work or not desirable to move the puback timeout communication into the status message IOT_CONNECT_STATUS_CB?

Hi Nik,
I did not add into IOT_CONNECT_STATUS_CB because from the name "IOT_CONNECT_STATUS_CB" it is clearly indicate that this callback is for reporting "connect status" only.

**Comment by alanlhc** [comment]
Hi Nik,

I did not add into IOT_CONNECT_STATUS_CB because from the name “IOT_CONNECT_STATUS_CB” it clearly indicate that this callback is for reporting “connect status” only.

Best regards,
Alan Low

**Comment by alanlhc** [comment]
Low, Alan would like to recall the message, "[avnet-iotconnect/iotc-nrf-sdk] Br wads fix 20210413 1428 (#2)".

**Comment by alanlhc** [comment]
Hi Nik,

I did not add into IOT_CONNECT_STATUS_CB because from the name “IOT_CONNECT_STATUS_CB” it clearly indicate that this callback is for reporting “connect status” only.


Best regards,
Alan Low

**Comment by alanlhc** [comment]
Hi Nik,

I did not change this. This is the original code.

Best regards,
Alan Low

**Comment by alanlhc** [comment]
Hi Nik,

Noted. Will remove the comment blocks.

Best regards,
Alan Low

**Comment by alanlhc** [comment]
Hi Nik,

What mqtt_disconnect() does:
1a) send “MQTT Disconnect notification” message to the broker to disconnect cleanly.
1b) close TCP connection.
1c) Fire MQTT event callback with MQTT_EVT_DISCONNECT.

What mqtt_abort() does:
2a) close TCP connection.
2b) Fire MQTT event callback with MQTT_EVT_DISCONNECT.

When LTE link is down, (1a) will not work if we call mqtt_disconnect().

Best regards,
Alan Low

**Comment by alanlhc** [comment]
> Would it be too much work or not desirable to move the puback timeout communication into the status message IOT_CONNECT_STATUS_CB?

Hi Nik,
I did not add into IOT_CONNECT_STATUS_CB because from the name "IOT_CONNECT_STATUS_CB" it is clearly indicate that this CB is for reporting 

> > Would it be too much work or not desirable to move the puback timeout communication into the status message IOT_CONNECT_STATUS_CB?
> 
> Hi Nik,
> I did not add into IOT_CONNECT_STATUS_CB because from the name "IOT_CONNECT_STATUS_CB" it is clearly indicate that this callback is for reporting "connect status" only.

**Comment by alanlhc** [comment]
Sorry I accidentally close it.

File: iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c (hunk 1 above)
Comment (alanlhc) [reply iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c:205]: Hi Nik,

I did not change this code. This is the original.

**Comment by alanlhc** [comment]
> Comment blocks that don't add value like "Function to find msg ID from the list." should be removed because the function name msg_wait_puback_list_find makes is obvious that it's supposed to do.

Hi Nik,

Noted.

File: iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c (hunk 2 above)
Comment (alanlhc) [reply iotconnect-sdk/nrf-layer-lib/src/iotconnect_mqtt.c:628]: Hi Nik,

What mqtt_disconnect() does:
1a) send “MQTT Disconnect notification” message to the broker to disconnect cleanly.
1b) close TCP connection.
1c) Fire MQTT event callback with MQTT_EVT_DISCONNECT.

What mqtt_abort() does:
2a) close TCP connection.
2b) Fire MQTT event callback with MQTT_EVT_DISCONNECT.

When LTE link is down, (1a) will not work if we call mqtt_disconnect().

**Pull request closed without merging**